### ISSUE TITLE
`update_function` removed from `spatial_smooth` method

### ISSUE THREAD

**Issue by brechmos-stsci** [radio-astro-tools/spectral-cube#533]
We are using the `update_function` parameter of `def spatial_smooth()` to get updates on the smoothing as it progresses.  It looks like the refactor of the smoothing in 0.4.4 removed this ability, but not sure if that was a decision or an oops. 

The traceback is:
```
---------------------------------------------------------------------------                                                                      
TypeError                                 Traceback (most recent call last)                                                                      
<ipython-input-21-db517e558904> in <module>
----> 1 cube.spatial_smooth(k, update_function=lambda x: print('here'))

/opt/miniconda3/envs/cubeviz-dev/lib/python3.6/site-packages/spectral_cube/spectral_cube.py in spatial_smooth(self, kernel, convolve, **kwargs)  
   2604
   2605         newcube = self.apply_function_parallel_spatial(_gsmooth_image,                                                                   
-> 2606                                                        **kwargs)
   2607
   2608         return newcube

/opt/miniconda3/envs/cubeviz-dev/lib/python3.6/site-packages/spectral_cube/spectral_cube.py in apply_function_parallel_spatial(self, function, num
_cores, verbose, use_memmap, parallel, **kwargs)
   2789                                                   num_cores=num_cores,                                                                   
   2790                                                   use_memmap=use_memmap,                                                                 
-> 2791                                                   **kwargs)
   2792
   2793     def apply_function_parallel_spectral(self,

/opt/miniconda3/envs/cubeviz-dev/lib/python3.6/site-packages/spectral_cube/spectral_cube.py in _apply_function_parallel_base(self, iteration_data,
 function, applicator, num_cores, verbose, use_memmap, parallel, memmap_dir, **kwargs)                                                           
   2726
   2727             for arg in iteration_data:
-> 2728                 applicator(arg, outcube, function, **kwargs)
   2729                 pbu()
   2730

/opt/miniconda3/envs/cubeviz-dev/lib/python3.6/site-packages/spectral_cube/spectral_cube.py in _apply_spatial_function(arguments, outcube, functio
n, **kwargs)
    118
    119     if np.any(includemask):
--> 120         outcube[ii, :, :] = function(img, **kwargs)
    121     else:
    122         outcube[ii, :, :] = img

/opt/miniconda3/envs/cubeviz-dev/lib/python3.6/site-packages/spectral_cube/spectral_cube.py in _gsmooth_image(img, **kwargs)                     
   2601             Helper function to smooth an image
   2602             """
-> 2603             return convolve(img, kernel, normalize_kernel=True, **kwargs)                                                                
   2604
   2605         newcube = self.apply_function_parallel_spatial(_gsmooth_image,                                                                   

TypeError: convolve() got an unexpected keyword argument 'update_function'                                                                
```

To replicate:
```
>>> cube = spectral_cube.SpectralCube.read(filename)
>>> k = astropy.convolution.Gaussian2DKernel(3)    
>>> cube.spatial_smooth(k, update_function=lambda x: print('here'))
```

Essentially, previously the `update_function` parameter was consumed in the definition of the `spatial_smooth` method:
```
    def spatial_smooth(self, kernel,
                       #numcores=None,
                       convolve=convolution.convolve,
                       update_function=None,
                       **kwargs):
```

and now it isn't and it is getting passed through to the `**kwargs` of astropy.convolution.convolve
```
 convolve(img, kernel, normalize_kernel=True, **kwargs)
```
which doesn't take that as a parameter. 

I believe this fell out in the https://github.com/radio-astro-tools/spectral-cube/commit/404ab98f1745f5f732028c0adb55c505fbbdcbce commit.

**Comment by brechmos-stsci** [comment]
Also, for the ` def spatial_smooth_median(self, ksize, update_function=None, **kwargs):` it looks like the `update_function` parameter is there, but not actually used. (might be mis-understanding this one though). Anyway, might need to check both `def spatial_smooth` and `def spatial_smooth_median`.

**Comment by keflavich** [comment]
Thanks, this is likely an oversight from the refactor... we now moved all the smoothing stuff into a common `apply_function_parallel_spatial` tool.  

I'm glad you're using it!  We'll add a regression test too.

**Comment by keflavich** [comment]
General question from the cubeviz side-  are you guys pinning to release versions?  Do you have any test jobs that run against dev?  It would help us to know if/when we're likely to break things.  0.4.4 was a rather large change from 0.4.3, but mostly because we neglected making releases over the last year.

**Comment by brechmos-stsci** [comment]
Typically we are not pinning except from the bottom.  We just added a pin temporarily to the top for this one as it is a breaking change and we are planning a cubeviz release in the next week or so.  :/ 

We don't have a job that runs against master but we do use the latest release.  We could/should add one in there for master as we do use spectral-cube for all the actual processing.

**Comment by keflavich** [comment]
@brechmos-stsci Could you say a little more about what you need the `update_function` for?  It turns out this was a pretty fundamental change in that we switched from using a `for` loop to loop over the spectral channels to using a `map` operation, which means that in many cases there is no longer a guarantee the operation will be done in linear order.

It also means that, because of how we're applying the function (smooth, etc), we don't have a trivial way to add this callback.

**Comment by keflavich** [comment]
I see, you're using it here:
https://github.com/spacetelescope/cubeviz/blob/master/cubeviz/tools/smoothing.py#L387

I... think... we can do something about this.  It will take some thought and I'll ask for your review on the PR.

**Comment by keflavich** [comment]
@brechmos-stsci could you test against #534?  It's my attempted fix, but it might be somewhat questionable.... I'm not sure I'm using a supported API in `joblib`.

**Comment by brechmos-stsci** [comment]
@keflavich I think this can be closed now as #534 was merged.  Unless you wanted to keep it open for a different fix (?).
